src/compiler/crystal/semantic/type_merge.cr

```diff
@@ -111,25 +129,54 @@ module Crystal
       # Nothing to do
     end
 
-    def combined_union_of(types : Array)
+    def combined_union_of(types : Array, merge_to_parent = true)
       case types.size
       when 0
         nil
       when 1
         types.first
       else
-        combined_types = type_combine types
+        combined_types = type_combine types, merge_to_parent: merge_to_parent
         union_of combined_types
       end
     end
 
-    def type_combine(types)
+    # If merge to parent is true, types will be merged to common ancestors.
+    # If false, types will only be merged to a common ancestor if any of those
+    # types is the common ancestor. For example:
+    #
+    # ```
+    # class Base; end
+    #
+    # class One < Base; end
+    #
+    # class Two < Base; end
+    # ```
+    #
+    # Merging One | Two:
+    # - merge_to_parent = true: Base+
+    # - merge_to_parent = false: One | Two
+    #
+    # Merging Base | One:
+    # - merge_to_parent = true: Base
+    # - merge_to_parent = false: Base+
+    def type_combine(types, merge_to_parent = true)
       all_types = [types.shift] of Type
 
+      # Sort by depth so that types deep in the hierarchy will combine with
+      # types we find first.
+      if merge_to_parent == false
+        types = types.sort_by!(&.depth)
```

Comment:
```suggestion
        types = types.sort_by(&.depth)
```

Reply:
Oh, it should just sort in place. Thanks!

Reply:
I don't think so, since you shouldn't modify the method argument.

Reply:
I do think so, because I know it's safe to mutate the argument and I don't want to allocate more memory.